### ISSUE TITLE
Runtime sized arrays of accelerationStructureNV's

### ISSUE THREAD

**Issue by RikoOphorst** [KhronosGroup/glslang#1766]
glslang produces this error:

> array must be redeclared with a size before being indexed with a variable

when asked to compile

`uniform accelerationStructureNV myAccelerationStructures[];`

I haven't looked very deeply into this, but from my understanding GL_EXT_nonuniform_qualifier should allow this type of uniform array if the array is indexed with dynamic expressions such as:

```
uniform metaData{ uint accelerationStructureIndex; };
uniform accelerationStructureNV myAccelerationStructures[];

void myTraceFunc()
{
	traceNV(myAccelerationStructures[accelerationStructureIndex], 0, 0xFF, 1, 0, 1, vec3(0,0,0), 0.0001f, vec3(1, 0, 0), 1000.0f, 1);
}
```

My suspicion is that this behavior is caused by this block of code: 
https://github.com/KhronosGroup/glslang/blob/c2c5a3d973310b2c9806d34c788fa82508128efd/glslang/MachineIndependent/ParseHelper.cpp#L3897

It seems this check does not take into account that accelerationStructureNV's also can be arrayed. Or at least, I _think_ they can be put in run time sized arrays.

Is this a glslang bug? Or is this intended behavior? Thanks!

**Comment by johnkslang** [comment]
I don't see any examples of this type getting an array?

@jeffbolznv ?

**Comment by RikoOphorst** [comment]
@johnkslang 

So the GLSL_NV_ray_tracing spec states this:

> accelerationStructureNV is an opaque type representing a structure used during
>     ray tracing to accelerate queries of intersections of rays with scene geometry.
>     It is declared and behaves like above described opaque types. When aggregated
>     into arrays within a shader, accelerationStructureNV can only be indexed with
>     a dynamically uniform integral expression, otherwise results are undefined.

This implies that it is possible to array this type. Regardless, doing something like this _does_ work:

```
uniform metaData{ uint accelerationStructureIndex; };
uniform accelerationStructureNV myAccelerationStructures[2500];

void myTraceFunc()
{
	traceNV(myAccelerationStructures[accelerationStructureIndex], 0, 0xFF, 1, 0, 1, vec3(0,0,0), 0.0001f, vec3(1, 0, 0), 1000.0f, 1);
}
```

However it'd be very desirable to let it be a runtime sized array depending on the amount of descriptors that you bound to the binding.

**Comment by jeffbolznv** [comment]
I discussed this with Eric and we agreed this ought to be allowed. I'll take care of it.

**Comment by jeffbolznv** [comment]
Hmm, this is a bit more subtle than I initially realized. There are a couple places the Vulkan spec says dynamic indices are not allowed (interfaces chapter) and runtime array is not allowed (spirv environment appendix). Fixing these in a way that's consistent with everything else would need new SPIR-V capabilities and new Vulkan feature bits. So I need to discuss with others how we want to deal with this.

**Comment by RikoOphorst** [comment]
@jeffbolznv Thanks for looking into it. Though, a lot of this behavior is made possible by the duo of VK_EXT_descriptor_indexing and GL_EXT_nonuniform_qualifier. For all "normal" types like SSBO's, UBO's, images, samplers, etc it is possible to get these types to be runtime sized arrays - if using those extensions. My assumption is that a lot of these feature bits & spir-v capabilities are already (sort of) available in these extensions. Though, I of course don't know much about the underlying driver & hardware stuff you guys need to do to get this working.

Could you post an update once you know more? Thanks a lot!

**Comment by jeffbolznv** [comment]
The plan is to allow this without requiring and new features/capabilities. I'll put together the glslang change to fix this. I was wrong about "runtime array is not allowed (spirv environment appendix)" since acceleration structures go in the UniformConstant storage class and that's already allowed, so that hopefully means no spirv-val change will be necessary either.

**Comment by jeffbolznv** [comment]
I pushed a PR to allow this in glslang. I've also made a small (non-public) PR against the Vulkan spec to clarify that dynamic indexing is supported in the interfaces.txt chapter, which will hopefully be released in a couple weeks.

**Comment by oddhack** [comment]
The Vulkan change mentioned by @jeffbolznv above has been published in the 1.1.111 Vulkan-Docs spec update.
